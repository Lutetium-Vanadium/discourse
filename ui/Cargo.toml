[package]
name = "ui"
version = "0.0.1"
authors = ["Lutetium Vanadium"]
edition = "2018"

[dependencies]
bitflags = "1.2"
lazy_static = "1.4"
unicode-segmentation = "1.7"

async-trait = { version = "0.1", optional = true }
futures = { version = "0.3", optional = true }

crossterm = { version = "0.19", optional = true }
termion = { version = "1.5", optional = true }
# TODO: add all of these
# easycurses = { version = "0.12", optional = true }
# pancurses = { version = "0.16", optional = true, features = ["win32a"] }

# The following dependencies are renamed in the form `<dep-name>-dep` to allow
# a feature name with the same name to be present. This is necessary since
# these features are required to enable other dependencies
smol-dep = { package = "smol", version = "1.2", optional = true }
tokio-dep = { package = "tokio", version = "1.5", optional = true, features = ["fs", "rt"] }
async-std-dep = { package = "async-std", version = "1.9", optional = true, features = ["unstable"] }

[features]
default = []
# curses = ["easycurses", "pancurses"]
tokio = ["tokio-dep", "async-trait", "futures", "anes"]
smol = ["smol-dep", "async-trait", "futures", "anes"]
async-std = ["async-std-dep", "async-trait", "futures", "anes"]

[target.'cfg(unix)'.dependencies]
anes = { version = "0.1", optional = true, features = ["parser"] }
