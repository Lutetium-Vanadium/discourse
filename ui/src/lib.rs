#![deny(missing_docs, rust_2018_idioms)]
//! A widget based cli ui rendering library
use std::{convert::TryFrom, io, sync::Mutex};

use crossterm::{
    cursor, event, execute, queue,
    style::{Colorize, Print, PrintStyledContent, Styler},
    terminal,
};

pub use widget::Widget;

/// In build widgets
pub mod widgets {
    pub use crate::char_input::CharInput;
    pub use crate::list::{List, ListPicker};
    pub use crate::string_input::StringInput;

    /// The default type for filter_map_char in [`StringInput`] and [`CharInput`]
    pub type FilterMapChar = fn(char) -> Option<char>;

    /// Character filter that lets every character through
    pub fn no_filter(c: char) -> Option<char> {
        Some(c)
    }
}

mod char_input;
mod list;
mod string_input;
mod widget;

/// Returned by [`Prompt::validate`]
pub enum Validation {
    /// If the prompt is ready to finish.
    Finish,
    /// If the state is valid, but the prompt should still persist.
    /// Unlike returning an Err, this will not print anything unique, and is a way for the prompt to
    /// say that it internally has processed the `Enter` key, but is not complete.
    Continue,
}

/// This trait should be implemented by all 'root' widgets.
///
/// It provides the functionality specifically required only by the main controlling widget. For the
/// trait required for general rendering to terminal, see [`Widget`].
pub trait Prompt: Widget {
    /// The error type returned by validate. It **must** be only one line long.
    type ValidateErr: std::fmt::Display;

    /// The output type returned by [`Input::run`]
    type Output;

    /// The main prompt text. It is printed in bold.
    fn prompt(&self) -> &str;
    /// The hint text. If a hint is there, it is printed in dark grey.
    fn hint(&self) -> Option<&str> {
        None
    }

    /// Determine whether the prompt state is ready to be submitted. It is called whenever the use
    /// presses the enter key.
    #[allow(unused_variables)]
    fn validate(&mut self) -> Result<Validation, Self::ValidateErr> {
        Ok(Validation::Finish)
    }
    /// The value to return from [`Input::run`]. This will only be called after
    /// [`validate`](Prompt::validate), if validate returns `Ok(true)`
    fn finish(self) -> Self::Output;

    /// The prompt has some default value that can be returned.
    fn has_default(&self) -> bool;
    /// The default value to be returned. It will only be called when has_default is true and the
    /// user presses escape.
    fn finish_default(self) -> Self::Output
    where
        Self: Sized,
    {
        unreachable!();
    }
}

/// The ui runner. It renders and processes events with the help of a type that implements [`Prompt`]
///
/// See [`run`](Input::run) for more information
pub struct Input<P> {
    prompt: P,
    terminal_h: u16,
    terminal_w: u16,
    base_row: u16,
    base_col: u16,
    error_row: Option<u16>,
    hide_cursor: bool,
}

impl<P: Prompt> Input<P> {
    fn adjust_scrollback<W: io::Write>(
        &self,
        height: usize,
        stdout: &mut W,
    ) -> crossterm::Result<u16> {
        let th = self.terminal_h as usize;

        let mut base_row = self.base_row;

        if self.base_row as usize >= th - height {
            let dist = (self.base_row as usize + height - th + 1) as u16;
            base_row -= dist;
            queue!(stdout, terminal::ScrollUp(dist), cursor::MoveUp(dist))?;
        }

        Ok(base_row)
    }

    fn set_cursor_pos<W: io::Write>(&self, stdout: &mut W) -> crossterm::Result<()> {
        let (dcw, dch) = self.prompt.cursor_pos(self.base_col);
        execute!(stdout, cursor::MoveTo(dcw, self.base_row + dch))
    }

    fn render<W: io::Write>(&mut self, stdout: &mut W) -> crossterm::Result<()> {
        let height = self.prompt.height();
        self.base_row = self.adjust_scrollback(height, stdout)?;
        self.clear(self.base_col, stdout)?;
        queue!(stdout, cursor::MoveTo(self.base_col, self.base_row))?;

        self.prompt
            .render((self.terminal_w - self.base_col) as usize, stdout)?;

        self.set_cursor_pos(stdout)
    }

    fn clear<W: io::Write>(&self, prompt_len: u16, stdout: &mut W) -> crossterm::Result<()> {
        if self.base_row + 1 < self.terminal_h {
            queue!(
                stdout,
                cursor::MoveTo(0, self.base_row + 1),
                terminal::Clear(terminal::ClearType::FromCursorDown),
            )?;
        }

        queue!(
            stdout,
            cursor::MoveTo(prompt_len, self.base_row),
            terminal::Clear(terminal::ClearType::UntilNewLine),
        )
    }

    #[inline]
    fn finish<W: io::Write>(
        self,
        pressed_enter: bool,
        prompt_len: u16,
        stdout: &mut W,
    ) -> crossterm::Result<P::Output> {
        self.clear(prompt_len, stdout)?;
        stdout.flush()?;
        if pressed_enter {
            Ok(self.prompt.finish())
        } else {
            Ok(self.prompt.finish_default())
        }
    }

    /// Run the ui on the given writer. It will return when the user presses `Enter` or `Escape`
    /// based on the [`Prompt`] implementation.
    pub fn run<W: io::Write>(mut self, stdout: &mut W) -> crossterm::Result<P::Output> {
        let (tw, th) = terminal::size()?;
        self.terminal_h = th;
        self.terminal_w = tw;

        let prompt = self.prompt.prompt();
        let prompt_len = u16::try_from(prompt.chars().count() + 3).expect("really big prompt");

        let _raw = RawMode::enable()?;
        if self.hide_cursor {
            queue!(stdout, cursor::Hide)?;
        };

        let height = self.prompt.height();
        self.base_row = cursor::position()?.1;
        self.base_row = self.adjust_scrollback(height, stdout)?;

        queue!(
            stdout,
            PrintStyledContent("? ".green()),
            PrintStyledContent(prompt.bold()),
            Print(' '),
        )?;

        let hint_len = match self.prompt.hint() {
            Some(hint) => {
                queue!(stdout, PrintStyledContent(hint.dark_grey()), Print(' '))?;
                u16::try_from(hint.chars().count() + 1).expect("really big prompt")
            }
            None => 0,
        };

        self.base_col = prompt_len + hint_len;

        self.render(stdout)?;

        loop {
            match event::read()? {
                event::Event::Resize(tw, th) => {
                    self.terminal_w = tw;
                    self.terminal_h = th;
                }

                event::Event::Key(e) => {
                    if let Some(error_row) = self.error_row.take() {
                        let pos = cursor::position()?;
                        queue!(
                            stdout,
                            cursor::MoveTo(0, error_row),
                            terminal::Clear(terminal::ClearType::CurrentLine),
                            cursor::MoveTo(pos.0, pos.1)
                        )?;
                    }

                    let key_handled = match e.code {
                        event::KeyCode::Char('c')
                            if e.modifiers.contains(event::KeyModifiers::CONTROL) =>
                        {
                            queue!(
                                stdout,
                                cursor::MoveTo(0, self.base_row + self.prompt.height() as u16)
                            )?;
                            drop(_raw);
                            if self.hide_cursor {
                                queue!(stdout, cursor::Show)?;
                            }
                            exit()
                        }
                        event::KeyCode::Null => {
                            queue!(
                                stdout,
                                cursor::MoveTo(0, self.base_row + self.prompt.height() as u16)
                            )?;
                            drop(_raw);
                            if self.hide_cursor {
                                queue!(stdout, cursor::Show)?;
                            }
                            exit()
                        }
                        event::KeyCode::Esc if self.prompt.has_default() => {
                            return self.finish(false, prompt_len, stdout);
                        }

                        event::KeyCode::Enter => match self.prompt.validate() {
                            Ok(Validation::Finish) => {
                                return self.finish(true, prompt_len, stdout);
                            }
                            Ok(Validation::Continue) => true,
                            Err(e) => {
                                let height = self.prompt.height() + 1;
                                self.base_row = self.adjust_scrollback(height, stdout)?;

                                queue!(stdout, cursor::MoveTo(0, self.base_row + height as u16))?;
                                write!(stdout, "{} {}", ">>".dark_red(), e)?;

                                self.error_row = Some(self.base_row + height as u16);

                                self.set_cursor_pos(stdout)?;

                                continue;
                            }
                        },
                        _ => self.prompt.handle_key(e),
                    };

                    if key_handled {
                        self.render(stdout)?;
                    }
                }
                _ => {}
            }
        }
    }
}

impl<P> Input<P> {
    /// Creates a new Input
    pub fn new(prompt: P) -> Self {
        Self {
            prompt,
            base_row: 0,
            base_col: 0,
            terminal_h: 0,
            terminal_w: 0,
            error_row: None,
            hide_cursor: false,
        }
    }

    /// Hides the cursor while running the input
    pub fn hide_cursor(mut self) -> Self {
        self.hide_cursor = true;
        self
    }
}

lazy_static::lazy_static! {
    static ref EXIT_HANDLER: Mutex<fn() -> !> = Mutex::new(default_exit);
}

/// Sets the exit handler to call when CTRL+C is received
pub fn set_exit_handler(handler: fn() -> !) {
    *EXIT_HANDLER.lock().unwrap() = handler;
}

fn default_exit() -> ! {
    std::process::exit(130);
}

fn exit() -> ! {
    match EXIT_HANDLER.lock() {
        Ok(exit) => exit(),
        Err(_) => default_exit(),
    }
}

/// Simple helper to make sure if the code panics in between, raw mode is disabled
struct RawMode {
    _private: (),
}

impl RawMode {
    /// Enable raw mode for the terminal
    pub fn enable() -> crossterm::Result<Self> {
        terminal::enable_raw_mode()?;
        Ok(Self { _private: () })
    }
}

impl Drop for RawMode {
    fn drop(&mut self) {
        let _ = terminal::disable_raw_mode();
    }
}
