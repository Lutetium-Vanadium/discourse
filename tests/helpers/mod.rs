use std::io;

use ui::{backend, error, style};

pub struct SnapshotOnFlushBackend {
    backend: backend::TestBackend,
}

impl SnapshotOnFlushBackend {
    #[allow(dead_code)]
    pub fn new(size: backend::Size) -> Self {
        Self {
            backend: backend::TestBackend::new(size),
        }
    }

    #[allow(dead_code)]
    pub fn new_with_layout(size: backend::Size, layout: ui::layout::Layout) -> Self {
        Self {
            backend: backend::TestBackend::new_with_layout(size, layout),
        }
    }
}

impl io::Write for SnapshotOnFlushBackend {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.backend.write(buf)
    }

    fn flush(&mut self) -> io::Result<()> {
        ui::assert_backend_snapshot!(self.backend);
        Ok(())
    }
}

impl backend::Backend for SnapshotOnFlushBackend {
    fn enable_raw_mode(&mut self) -> error::Result<()> {
        self.backend.enable_raw_mode()
    }

    fn disable_raw_mode(&mut self) -> error::Result<()> {
        self.backend.disable_raw_mode()
    }

    fn hide_cursor(&mut self) -> error::Result<()> {
        self.backend.hide_cursor()
    }

    fn show_cursor(&mut self) -> error::Result<()> {
        self.backend.show_cursor()
    }

    fn get_cursor_pos(&mut self) -> error::Result<(u16, u16)> {
        self.backend.get_cursor_pos()
    }

    fn move_cursor_to(&mut self, x: u16, y: u16) -> error::Result<()> {
        self.backend.move_cursor_to(x, y)
    }

    fn move_cursor(&mut self, direction: backend::MoveDirection) -> error::Result<()> {
        self.backend.move_cursor(direction)
    }

    fn scroll(&mut self, dist: i16) -> error::Result<()> {
        self.backend.scroll(dist)
    }

    fn set_attributes(&mut self, attributes: style::Attributes) -> error::Result<()> {
        self.backend.set_attributes(attributes)
    }

    fn set_fg(&mut self, color: style::Color) -> error::Result<()> {
        self.backend.set_fg(color)
    }

    fn set_bg(&mut self, color: style::Color) -> error::Result<()> {
        self.backend.set_bg(color)
    }

    fn clear(&mut self, clear_type: backend::ClearType) -> error::Result<()> {
        self.backend.clear(clear_type)
    }

    fn size(&self) -> error::Result<backend::Size> {
        self.backend.size()
    }
}
