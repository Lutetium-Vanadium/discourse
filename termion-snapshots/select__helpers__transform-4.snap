---
source: tests/helpers/mod.rs
expression: self.backend

---
┌──────────────────────────────────────────────────┐
│[38;5;10m✔[39m [1mcheckbox[22m [38;5;8m·[39m [38;5;5m8: Choice 8[39m                          │
│[38;5;0m[48;5;7m [39m[49m                                                 │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
└──────────────────────────────────────────────────┘
